src/fsharp/TypeChecker.fs

```diff
@@ -9278,38 +9293,70 @@ and TcNameOfExpr cenv env tpenv (synArg: SynExpr) =
+            let resolvesAsExpr =
+                match nameResolutionResult with 
+                | Result ((item, _, _, _) as res) 
+                    when 
+                         (match item with 
+                          | Item.Types _ 
+                          | Item.DelegateCtor _
+                          | Item.CtorGroup _
+                          | Item.FakeInterfaceCtor _ -> false
+                          | _ -> true) -> 
+                    let overallTy = match overallTyOpt with None -> NewInferenceType() | Some t -> t 
+                    let _, _ = TcItemThen cenv overallTy env tpenv res delayed
+                    true
+                | _ ->
                     false
-            if resolvedToModuleOrNamespaceName then result else
+            if resolvesAsExpr then result else
 
+            // If it's not an expression then try to resolve it as a type name
             let resolvedToTypeName = 
                 if (match delayed with [DelayedTypeApp _] | [] -> true | _ -> false) then
                     let (TypeNameResolutionInfo(_, staticArgsInfo)) = GetLongIdentTypeNameInfo delayed
                     match ResolveTypeLongIdent cenv.tcSink cenv.nameResolver ItemOccurence.UseInAttribute OpenQualified env.eNameResEnv ad longId staticArgsInfo PermitDirectReferenceToGeneratedType.No with
-                    | Result tcref when (match delayed with [DelayedTypeApp _] | [] -> true | _ -> false) && IsEntityAccessible cenv.amap m ad tcref -> 
+                    | Result tcref when IsEntityAccessible cenv.amap m ad tcref -> 
+                        match delayed with 
+                        | [DelayedTypeApp (tyargs, _, mExprAndTypeArgs)] -> 
+                            TcTypeApp cenv NewTyparsOK CheckCxs ItemOccurence.UseInType env tpenv mExprAndTypeArgs tcref [] tyargs |> ignore
+                        | _ -> ()
                         true // resolved to a type name, done with checks
                     | _ -> 
                         false
                 else
                     false
             if resolvedToTypeName then result else
 
-            let overallTy = match overallTyOpt with None -> NewInferenceType() | Some t -> t 
-            
-            // This will raise an error if resolution doesn't succeed
-            let _, _ = TcLongIdentThen cenv overallTy env tpenv lidd delayed
-            result // checked as an expression, done with checks
+            // If it's not an expression or type name then resolve it as a module
+            let resolvedToModuleOrNamespaceName =
+                if delayed.IsEmpty then
+                    let id,rest = List.headAndTail longId
+                    match ResolveLongIndentAsModuleOrNamespaceOrStaticClass cenv.tcSink ResultCollectionSettings.AllResults cenv.amap m false true OpenQualified env.eNameResEnv ad id rest true with 
```

Comment:
Note that with open type declarations coming in this call may no longer do what is assumed, see here: https://github.com/dotnet/fsharp/pull/9513/files#diff-dbb9e59b8ac1e1e04e1b00282a91ec97R2192

Does that have any impact? I would expect not...